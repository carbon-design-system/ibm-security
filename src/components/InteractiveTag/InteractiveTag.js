/**
 * @file Interactive tag.
 * @copyright IBM Security 2019 - 2020
 */

import Close16 from '@carbon/icons-react/lib/close/16';

import classnames from 'classnames';
import { bool, func, string } from 'prop-types';
import React from 'react';

import { getComponentNamespace } from '../../globals/namespace/index';

import Icon from '../Icon';
import Tag from '../Tag';

export const namespace = getComponentNamespace('tag--interactive');

/**
 * Interactive tag component.
 */
const InteractiveTag = ({
  children,
  className,
  isSelected,
  onRemove,
  removable,
  removeBtnLabel,
  ...other
}) => (
  <Tag
    className={classnames(namespace, className, {
      [`${namespace}--default`]: !isSelected,
      [`${namespace}--removable`]: removable,
      [`${namespace}--selected`]: isSelected,
    })}
    filter={false}
    {...other}
  >
    {children}
    {removable && (
      <button
        aria-label={removeBtnLabel}
        className={`${namespace}__button`}
        onClick={onRemove}
      >
        <Icon className={`${namespace}__button__icon`} renderIcon={Close16} />
      </button>
    )}
  </Tag>
);

InteractiveTag.defaultProps = {
  isSelected: false,
  onRemove: null,
  removable: false,
  removeBtnLabel: 'Remove',
  type: 'gray',
};

const { propTypes } = Tag;

InteractiveTag.propTypes = {
  ...propTypes,
  type: propTypes.type,

  /** @type {boolean} Determines if the tag is selected. */
  isSelected: bool,

  /** @type {function} Callback function to remove the tag. */
  onRemove: func,

  /** @type {boolean} Set whether the tag is removable or not. */
  removable: bool,

  /** @type {string} Label for the button used to remove the tag. */
  removeBtnLabel: string,
};

export default InteractiveTag;
