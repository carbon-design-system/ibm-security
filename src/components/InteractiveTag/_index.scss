////
/// Interactive tag component.
/// @group tag--interactive
/// @copyright IBM Security 2019 - 2020
////

@import '../Tag/index';

@import '../Component/mixins';
@import 'mixins';

@include security--component($name: $tag--interactive__name) {
  @include tag--interactive;
}
