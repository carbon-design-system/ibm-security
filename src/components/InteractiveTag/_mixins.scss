////
/// Interactive tag mixins.
/// @group tag--interactive
/// @copyright IBM Security 2019 - 2021
////

@import '@carbon/themes/scss/tokens';
@import '../../globals/namespace/index';

/// Name.
/// @type String
$tag--interactive__name: tag--interactive;

/// Namespace.
/// @type String
$tag--interactive__namespace: get-component-namespace(
  $component-name: $tag--interactive__name,
);

@mixin tag--interactive {
  &--default {
    background-color: $active-ui;
    color: $text-01;
  }

  > .#{$prefix}--tag__close-icon:hover {
    background-color: inherit;
  }
}
