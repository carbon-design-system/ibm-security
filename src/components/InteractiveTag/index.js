/**
 * @file Interactive tag entry point.
 * @copyright IBM Security 2019
 */

export default from './InteractiveTag';
