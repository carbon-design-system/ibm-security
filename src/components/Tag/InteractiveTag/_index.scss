////
/// Interactive tag component.
/// @group tag--interactive
/// @copyright IBM Security 2019
////

@import '../../Component/mixins';
@import 'mixins';

@include security--component($name: $tag--interactive__name) {
  @include tag--interactive;
}
