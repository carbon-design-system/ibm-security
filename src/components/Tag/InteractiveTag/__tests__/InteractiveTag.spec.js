/**
 * @file Interactive tag tests.
 * @copyright IBM Security 2019
 */

import { render } from '@testing-library/react';
import React from 'react';
import userEvent from '@testing-library/user-event';

import { InteractiveTag } from '../../../..';
import { namespace } from '../InteractiveTag';

describe('InteractiveTag', () => {
  test('should have no Axe or DAP violations', async () => {
    const main = document.createElement('main');
    render(<InteractiveTag removable>test tag</InteractiveTag>, {
      // DAP requires a landmark '<main>' in the DOM:
      container: document.body.appendChild(main),
    });
    await expect(document.body).toHaveNoAxeViolations();
    await expect(document.body).toHaveNoDAPViolations('InteractiveTag');
  });

  test('should add close button with `aria-label` when `removable` is `true`', () => {
    const { queryByLabelText } = render(
      <InteractiveTag removeBtnLabel="test label" removable>
        test tag
      </InteractiveTag>
    );
    expect(queryByLabelText(/test label/i)).toBeVisible();
  });

  test('should add children as tag content', () => {
    const { queryByText } = render(<InteractiveTag>test tag</InteractiveTag>);
    expect(queryByText(/test tag/i)).toBeVisible();
  });

  test('should add a custom class', () => {
    const { getByText } = render(
      <InteractiveTag className="custom-class">test tag</InteractiveTag>
    );
    expect(getByText(/test tag/i)).toHaveClass('custom-class');
  });

  test('should add selected class when `isSelected` is `true`', () => {
    const { getByText } = render(
      <InteractiveTag isSelected>test tag</InteractiveTag>
    );
    expect(getByText(/test tag/i)).toHaveClass(`${namespace}--selected`);
  });

  test('should pass through extra props via spread attribute', () => {
    const { queryByTestId } = render(
      <InteractiveTag data-testid="test-id">test tag</InteractiveTag>
    );
    expect(queryByTestId('test-id')).toBeInTheDocument();
  });

  test('should invoke remove mock when remove button is clicked', () => {
    const onRemoveMock = jest.fn();
    const { getByLabelText } = render(
      <InteractiveTag
        onRemove={onRemoveMock}
        removable
        removeBtnLabel="test remove button"
      >
        test tag
      </InteractiveTag>
    );

    userEvent.click(getByLabelText(/test remove button/i));
    expect(onRemoveMock).toHaveBeenCalledTimes(1);
  });
});
