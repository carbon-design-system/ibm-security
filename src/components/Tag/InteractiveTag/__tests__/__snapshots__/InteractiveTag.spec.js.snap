// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InteractiveTag Rendering renders correctly 1`] = `
<Tag
  className="security--tag--interactive security--tag--interactive--default security--tag--interactive--removable"
  filter={false}
  type="gray"
>
  Label
  <button
    aria-label="Remove"
    className="security--tag--interactive__button"
    onClick={[MockFunction]}
  >
    <Icon
      className="security--tag--interactive__button__icon"
      height={null}
      path={null}
      renderIcon={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      size={20}
      viewBox="0 0 32 32"
      width={null}
    />
  </button>
</Tag>
`;

exports[`InteractiveTag Rendering renders the HTML of the node's subtree 1`] = `
<span
  class="bx--tag bx--tag--gray security--tag--interactive security--tag--interactive--default security--tag--interactive--removable"
>
  Label
  <button
    aria-label="Remove"
    class="security--tag--interactive__button"
  >
    <svg
      aria-hidden="true"
      class="security--icon security--tag--interactive__button__icon"
      focusable="false"
      height="16"
      preserveAspectRatio="xMidYMid meet"
      style="will-change:transform"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 4.7L11.3 4 8 7.3 4.7 4 4 4.7 7.3 8 4 11.3 4.7 12 8 8.7 11.3 12 12 11.3 8.7 8z"
      />
    </svg>
  </button>
</span>
`;

exports[`InteractiveTag Rendering renders the selected variation correctly 1`] = `
<Tag
  className="security--tag--interactive security--tag--interactive--removable security--tag--interactive--selected"
  filter={false}
  type="gray"
>
  Label
  <button
    aria-label="Remove"
    className="security--tag--interactive__button"
    onClick={[MockFunction]}
  >
    <Icon
      className="security--tag--interactive__button__icon"
      height={null}
      path={null}
      renderIcon={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      size={20}
      viewBox="0 0 32 32"
      width={null}
    />
  </button>
</Tag>
`;
